### PR TITLE
add riemann events sink

### LINKED ISSUE #1160: Add Riemann sink for events
It seems "events" does not support the Riemann sink, which would be useful for health check alerts, etc.

### PR DESCRIPTION
**Pull request by shmish111**
Basically a copy and paste of metrics sink. The values I have chosen to send to Riemann are an opinion of what I would find useful based on my current Riemann usage.

fix #1160 

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/heapster/1339)

### CODE REVIEW

**Comment by shmish111** [comment]
Completed CLA with wrong google account at first so getting `commit missing GitHub user` however not sure how to fix that now?

**Comment by piosz** [comment]
@k8s-bot ok to test

**Comment by piosz** [comment]
Did you try `git commit --amend`? In the worts case you can try to just create a new PR.

Also I'm not familiarized with Riemann at all, so we need to find someone to review it.

@erimatnor @jfoy @jsoriano @huangyuqi are you able to make the review?

**Comment by jsoriano** [comment]
Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks broke.

---

*[events/sinks/riemann/driver.go, line 1 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/heapster/1339#-KU4FfGssI7Z54xbwQa_:-KU4FfGssI7Z54xbwQaa:b-8k88q1) ([raw file](https://github.com/kubernetes/heapster/blob/7bbff88f78c79ebdf645eefeefd8e23b9b2f45dd/events/sinks/riemann/driver.go#L1)):*
> ```Go
> // Copyright 2014 Google Inc. All Rights Reserved.
> ```

2016?

---

*[events/sinks/riemann/driver.go, line 48 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/heapster/1339#-KU4CBAxg7RXOBWYmk70:-KU4CBAy4jUYMg0QjKXk:b-pp5xdp) ([raw file](https://github.com/kubernetes/heapster/blob/7bbff88f78c79ebdf645eefeefd8e23b9b2f45dd/events/sinks/riemann/driver.go#L48)):*
> ```Go
> }
> 
> type riemannConfig struct {
> ```

It'd be nice to create a `k8s.io/heapster/common/riemann` package to avoid duplicating code between events and metrics sinks.

---

*[events/sinks/riemann/driver.go, line 167 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/heapster/1339#-KU4F2yDcH9zKiM5I_uW:-KU4F2yDcH9zKiM5I_uX:b170wee) ([raw file](https://github.com/kubernetes/heapster/blob/7bbff88f78c79ebdf645eefeefd8e23b9b2f45dd/events/sinks/riemann/driver.go#L167)):*
> ```Go
> 				"component":        event.Source.Component,
> 			},
> 			Metric: riemannValue(event.Count),
> ```

Does this event count make sense as metric? (Just asking, I don't have a better proposal in any case :))

---

*[events/sinks/riemann/driver_test.go, line 60 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/heapster/1339#-KU45i0nLkXfVwJTyD1s:-KU45i0nLkXfVwJTyD1t:b-4bpj7z) ([raw file](https://github.com/kubernetes/heapster/blob/7bbff88f78c79ebdf645eefeefd8e23b9b2f45dd/events/sinks/riemann/driver_test.go#L60)):*
> ```Go
> }
> 
> // Returns a fake kafka sink.
> ```

kafka? :)

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/heapster/1339)*

**Comment by jsoriano** [comment]
@shmish111 please, update sinks documentation too

**Comment by shmish111** [comment]
Sorry for the delay. I've updated the PR by extracting common logic.

WRT the event metric, I think this makes sense when reading http://kubernetes.io/docs/api-reference/v1/definitions/#_v1_event however it is opinionated to some extent, no way round that really.

**Comment by shmish111** [comment]
Build failing dur to someone else's code 
`FAIL	k8s.io/heapster/metrics/sources/kubelet [build failed]`

**Comment by shmish111** [comment]
@piosz I don't get the problem with the git commits, they all seem fine, all assigned to me and my github picture is there.

@jsoriano can you review my changes please.

Thanks!

**Comment by shmish111** [comment]
@k8s-bot test this

**Comment by shmish111** [comment]
@jsoriano the e2e tests failed due to my documentation change, travis failed for another reason, I can't work out why cla thinks my commits are missing Github user, basically the PR looks a mess however unit tests all work on my local machine. Not sure if any of this is a problem in getting this PR merged, if so then I'm not sure what to do.

**Comment by piosz** [comment]
@k8s-bot test this

**Comment by piosz** [comment]
Who is reviewing this? Please let me know when this is LGTMed and ready for the review

**Comment by jsoriano** [comment]
Reviewed 5 of 5 files at r2.
Review status: 6 of 7 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[events/sinks/riemann/driver.go, line 32 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/heapster/1339#-KXI7ZjSWMQzAA5StZaW:-KXI7ZjSWMQzAA5StZaX:b-t4z4uv) ([raw file](https://github.com/kubernetes/heapster/blob/109a9bcbc3294ee6b28e82c3cbb93a53fc910efd/events/sinks/riemann/driver.go#L32)):*
> ```Go
> 
> // Abstracted for testing: this package works against any client that obeys the
> // interface contract exposed by the goryman Riemann client
> ```

This comment was for the interface you have moved to common, move also the comment please :)

---

*[events/sinks/riemann/driver.go, line 125 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/heapster/1339#-KXIB-1IiI1MAHajHvHX:-KXIAkADnIZivCFJzuKJ:b-koe765) ([raw file](https://github.com/kubernetes/heapster/blob/109a9bcbc3294ee6b28e82c3cbb93a53fc910efd/events/sinks/riemann/driver.go#L125)):*
> ```Go
> }
> 
> func (sink *RiemannSink) sendData(dataEvents []riemann_api.Event) {
> ```

`sendData` could be moved to common, it's the same for metrics and events, isn't it?

---

*[events/sinks/riemann/driver_test.go, line 61 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/heapster/1339#-KXI8D1K-lJthCrol-U8:-KXI8D1K-lJthCrol-U9:bub6pmw) ([raw file](https://github.com/kubernetes/heapster/blob/109a9bcbc3294ee6b28e82c3cbb93a53fc910efd/events/sinks/riemann/driver_test.go#L61)):*
> ```Go
> }
> 
> // Returns a fake kafka sink.
> ```

kafka still here :)

---

*[metrics/sinks/riemann/driver.go, line 31 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/heapster/1339#-KXIBOeEe4laSSAoJQ-L:-KXIBOeEe4laSSAoJQ-M:b-bi52jp) ([raw file](https://github.com/kubernetes/heapster/blob/109a9bcbc3294ee6b28e82c3cbb93a53fc910efd/metrics/sinks/riemann/driver.go#L31)):*
> ```Go
> // interface contract exposed by the goryman Riemann client
> 
> type RiemannSink struct {
> ```

This struct is defined three times, maybe a better abstraction in common is possible.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/heapster/1339)*

**Comment by jsoriano** [comment]
I have added some comments to the review. Thanks for your changes @shmish111 !

**Comment by piosz** [comment]
@shmish111 please sign the CLA

**Comment by piosz** [comment]
@k8s-bot test this

**Comment by shmish111** [comment]
@jsoriano I've made some changes based on the comments
@piosz I have signed the cla, the problem is that github somehow thinks that some commits aren't mine, I have no idea how to fix this as all the commits seem to be properly associated with my github account.

File: events/sinks/riemann/driver.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+	riemann_api "github.com/bigdatadev/goryman"
+	"github.com/golang/glog"
+	riemannCommon "k8s.io/heapster/common/riemann"
+	core "k8s.io/heapster/events/core"
+	kube_api "k8s.io/kubernetes/pkg/api"
+)
+
+// Abstracted for testing: this package works against any client that obeys the
+// interface contract exposed by the goryman Riemann client
```
Comment (jsoriano) [inline events/sinks/riemann/driver.go]: This comment doesn't make much sense here if we don't keep the interface we had before.

File: metrics/sinks/riemann/driver.go [hunk 2]
```diff
@@ -16,114 +16,56 @@ package riemann
...
-	"time"
-
 	riemann_api "github.com/bigdatadev/goryman"
 	"github.com/golang/glog"
+	riemannCommon "k8s.io/heapster/common/riemann"
 	"k8s.io/heapster/metrics/core"
 )
 
 // Abstracted for testing: this package works against any client that obeys the
 // interface contract exposed by the goryman Riemann client
```
Comment (jsoriano) [inline metrics/sinks/riemann/driver.go]: Same here, this comment was connected with the removed interface.

File: common/riemann/riemann.go [hunk 3]
```diff
@@ -107,3 +108,30 @@ func RiemannValue(value interface{}) interface{} {
 	}
 	return value
 }
+
+func SendData(client RiemannClient, dataEvents []riemann_api.Event) {
```
Comment (jsoriano) [inline common/riemann/riemann.go]: Better this way, thanks :)

**Comment by jsoriano** [comment]
@shmish111 are you using Riemann sinks? we were thinking about deprecating them if they are not being used, you can see the conversations about that in #1407 and #1419, we are waiting for your opinion.
Thanks!

**Comment by jsoriano** [comment]
Reviewed 1 of 1 files at r3, 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/heapster/1339)*

**Comment by jsoriano** [comment]
@shmish111 about the CLA, I don't know if it's related, but is it possible that you are signing it with a different email than the used in the commit messages?

**Comment by shmish111** [comment]
@jsoriano sorry, I've had github notifications turned off for a while and didn't check your comments. 

I'm sorry to admit that I am not using heapster any more as I am working on a new project that doesn't have access to Kube (yet). Also, as a big fan of Riemann I am surprised to find myself drawn in to the new Elastic beats components, despite not being nearly as nice to configure as Riemann and being written in go (puke) ;-) I find the overall new BELK stack very appealing from an operational point of view. I am currently using Riemann but I can see it fading away and I can also see monitoring heapster with beats more likely to be allowed in to my current organization. I feel sad to say it but I think I agree with deprecating some of the sinks, including Riemann. I feel that heapster beats would be popular though, does it exist?

**Comment by mcorbin** [comment]
Hi,

I am currently working on the RIemann metric sink (to use https://github.com/riemann/riemann-go-client instead of Goryman). I will probably do the PR today. I will then work on the event sink.
I don't think Riemann is fading away, and the next Riemann release will bring a lot of improvements. Ofc, it will never be as popular as elastic/beat ;)

**Comment by shmish111** [comment]
@mcorbin good to hear Riemann is not dead! Will your work take the place of this PR?

I am currently torn between beats etc and Riemann, give me some ammo to push Riemann, I can't find any docs about new features coming or anything like that!

**Comment by mcorbin** [comment]
@shmish111 i changed several things on my PR on the metric sink (#1562), but I will take inspiration from your code ;)

Regarding Riemann, the last release was the 3 month ago (changelog [here](https://github.com/riemann/riemann/releases/tag/0.2.12)). Since was added things like microseconds time resolution in the protocol (limited to the second currently), complete refactoring of the influxdb output (should support all influxdb functionalities), new outputs (Kafka, netuitive), various bugfixes... And more to come ;) If you have questions about Riemann (i can give you some use cases when i use Riemann in my company for example), you can also contact me directly ;)

**Comment by jsoriano** [comment]
I think this PR can be closed as #1591 was already merged.

**Comment by DirectXMan12** [comment]
closing as per above comment.

**Pull request closed without merging**